src/internal/async-map-parallel.mjs

```diff
@@ -0,0 +1,52 @@
+import CircularBuffer from './circular-buffer'
+
+class ParallelMapper {
+  constructor (iter, mapFn, concurrency) {
+    this.iter = iter
+    this.buffer = new CircularBuffer(concurrency)
+    this.mapFn = mapFn
+    this.done = false
+  }
+
+  next () {
+    const { iter, buffer, done, mapFn } = this
+    let promise
+
+    if (buffer.isFull() || done) {
+      promise = buffer.shift()
+
+      if (!done) {
+        // Keep the buffer full!
+        this.next()
+      }
+    } else {
+      promise = iter.next().then(item => {
+        if (item.done) {
+          this.done = true
+          return item
+        }
+
+        const parallelPromise = mapFn(item.value).then(mappedValue => {
+          return { value: { value: item.value, mappedValue }, done: false }
+        })
+
+        if (!buffer.isFull()) {
+          // Go on filling up the buffer in the background
+          this.next()
+        }
+
+        return parallelPromise
+      })
+
+      buffer.push(promise)
```

Comment:
This is a problem. It needs to be combined with line 35, or else we return an item *and* buffer it which doesn't make sense. However something else is wrong, because when I do that other code breaks.